### PR TITLE
[MOB-15997] Analytics app extension support

### PR DESCRIPTION
**Pull request by cdhoffmann**
#68 

Support for App Extensions. In order to support App Extensions we have used the @availability attribute to mark the AnalyticsHelper is unavailable because it uses the UIApplication.shared instance. Then we made the Analytics Extension class a base class and created two children classes. One which will be used by Apps, and one which will be used by App Extensions when registering.

### CODE REVIEW

File: AEPAnalytics/Sources/Analytics.swift [hunk 1]
```diff
@@ -555,7 +578,7 @@ public class Analytics: NSObject, Extension {
     ///     - trackData: Dictionary containing tracking data
     ///     - timestamp: timestamp to use for tracking
     ///     - Returns a `Dictionary` containing the vars data
-    private func processAnalyticsVars(trackData: [String: Any]?, timestamp: TimeInterval) -> [String: String] {
+    fileprivate func processAnalyticsVars(trackData: [String: Any]?, timestamp: TimeInterval) -> [String: String] {
```
Comment (kevinlind) [inline AEPAnalytics/Sources/Analytics.swift:581]: Why use `fileprivate` access here? I don't see any other calls to this method except from `AnalyticsBase.track`.

File: Documentation/AEPAnalytics.md [hunk 2]
```diff
@@ -83,8 +83,27 @@ func application(_ application: UIApplication, didFinishLaunchingWithOptions lau
 }
 ```
 
+### App Extension Support
 
+If you are using AEPAnalytics from within an App Extension, make sure that you register the `AnalyticsAppExt` class with MobileCore instead of the `Analytics` class.
```
Comment (kevinlind) [inline Documentation/AEPAnalytics.md:88]: "... make sure that you register the `AnalyticsAppExtension` class ..."

File: Documentation/AEPAnalytics.md [hunk 3]
```diff
@@ -83,8 +83,27 @@ func application(_ application: UIApplication, didFinishLaunchingWithOptions lau
...
 ```
 
+### App Extension Support
 
+If you are using AEPAnalytics from within an App Extension, make sure that you register the `AnalyticsAppExt` class with MobileCore instead of the `Analytics` class. 
 
+**Swift**
+
+```swift
+ MobileCore.registerExtensions([AnalyticsAppExt.self, Identity.self], {
```
Comment (kevinlind) [inline Documentation/AEPAnalytics.md:93]: `[AnalyticsAppExtension.self, Identity.self]`

**Review by kevinlind — COMMENTED** [review]
Are most of the new test cases for the AnalyticsAppExtension just copies of the Analytics tests? If so, could you point out where you made changes for the app extension tests?

File: AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                ]
+            ]
+        ]
+        let ruleEngineEvent = Event(name: "Rule event", type: EventType.rulesEngine, source: EventSource.responseContent, data: eventData)
+        mockRuntime.simulateComingEvent(event: ruleEngineEvent)
+        waitForProcessing()
+        
+        let expectedVars = [
+            "ce": "UTF-8",
+            "pev2" : "AMACTION:testActionName",
```
Comment (cdhoffmann) [inline AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift:43]: @kevinlind This is an example of the difference in the tests. Anywhere where we check for CP variable in the `Analytics` functional tests, we no longer have that variable from the AppExtension child class.

File: Podfile.lock [hunk 5]
```diff
@@ -1,46 +0,0 @@
-PODS:
```
Comment (kevinlind) [inline Podfile.lock:1]: Why delete the Podfile.lock? Was this a mistake?

File: AnalyticsSampleAppExt/FileProviderEnumerator.swift [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    var enumeratedItemIdentifier: NSFileProviderItemIdentifier
+
+    init(enumeratedItemIdentifier: NSFileProviderItemIdentifier) {
+        self.enumeratedItemIdentifier = enumeratedItemIdentifier
+        super.init()
+    }
+
+    func invalidate() {
+        // TODO: perform invalidation of server connection if necessary
```
Comment (kevinlind) [inline AnalyticsSampleAppExt/FileProviderEnumerator.swift:25]: There are a lot of TODOs in this sample app. Is this work you'll do in another PR or is this FileProviderExtension a sample taken from someplace else and the actual implementation isn't needed for testing Analytics in an app extension?

File: AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift (hunk 4 above)
Comment (kevinlind) [reply AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift:43]: As the test classes are very similar but with a few changes, it would be great to find a way to reuse the majority of the code. I'm not sure how to do that, however. Though, it is something that can be taken up in another PR.

File: AnalyticsSampleAppExt/FileProviderEnumerator.swift (hunk 6 above)
Comment (cdhoffmann) [reply AnalyticsSampleAppExt/FileProviderEnumerator.swift:25]: This is just Apple's template for a File Extension.

File: Podfile.lock (hunk 5 above)
Comment (cdhoffmann) [reply Podfile.lock:1]: I was getting a CI cocoapods error where it said it couldn't find 3.6.0 so i wanted to see if removing the podfile.lock would fix it. And it did.

File: AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift (hunk 4 above)
Comment (cdhoffmann) [reply AEPAnalytics/Tests/FunctionalTests/AnalyticsAppExtConsequenceTests.swift:43]: This is a good point. I can get this worked out in this PR or it might never get done.

File: Podfile.lock (hunk 5 above)
Comment (cdhoffmann) [reply Podfile.lock]: done

**Review by kevinlind — APPROVED** [review]
